opencog/atoms/core/ScopeLink.cc

```diff
@@ -473,6 +474,28 @@ Handle ScopeLink::substitute_vardecl(const Handle& vardecl,
 	return createLink(oset, t);
 }
 
+Handle ScopeLink::consume_ill_quotations() const
+{
+	Handle vardecl = get_vardecl();
+	const Variables& variables = get_variables();
+	HandleSeq nouts;
+	for (size_t i = (get_vardecl() ? 1 : 0); i < get_arity(); ++i) {
+		Handle nbody = consume_ill_quotations(variables, getOutgoingAtom(i));
+		nouts.push_back(nbody);
+		// If the new body has terms with free variables but no
+		// vardecl it means that some quotations are missing. Rather
+		// than adding them we set vardecl to an empty VariableList.
```

Comment:
This comment does not make sense to me. Free variables are never declared. That is what makes them free. It's the definition of "free variable".

Reply:
At this point in the algorithm they really are free while they should be quoted. This part is largely a hack, it wrongly removes quotations, and instead putting them back I remove the variables from the final variable declaration. It's a hack and should be replaced as soon as it is not longer good enough.

Reply:
OK. I don't quite get what's happening here, but that OK. Thanks.